mono/metadata/object.c

```diff
@@ -2084,6 +2084,10 @@ mono_class_create_runtime_vtable (MonoDomain *domain, MonoClass *klass, MonoErro
 				/* it's a pointer type: add check */
 				g_assert ((fklass->byval_arg.type == MONO_TYPE_PTR) || (fklass->byval_arg.type == MONO_TYPE_FNPTR));
 				*t = *(char *)data;
+				/* This is not needed by sgen, as it does not seem 
++				to need write barriers for uncollectable objects (like the vtables storing static 
++				fields), but it is needed for incremental boehm. */
+				mono_gc_wbarrier_generic_nostore(t);
```

Comment:
Let me look into this. Perhaps we should always be scanning the uncollectable regions of memory in Boehm but we are not.

Reply:
I guess we could, if we wanted to match sgen's apparent behavior here. But I don't see a reason to do that. It seems more efficient to rescan the uncollectable objects only when they have changed, like everything else.

Reply:
My main worry is making sure we don't miss any new places these are needed or any we didn't hit yet.

Reply:
So, I discovered most of these in Unity and mono through my test coverage. I have a mode where I run patch the GC functions to perform tracking of memory and then verify all references match set write barriers at the end of the run. I then made sure that RuntimeTests would pass without any unaccounted references using this setup. I am currently going through the same process for il2cpp.

The patching of GC functions is done at runtime by injecting a jmp instruction at the beginning of the functions. That way I can add the patching code to our development builds, and enable it through some environment variable. That should make it relatively easy to add configs to katana to run these tests, as is should be possible to do this by just adding a new katana config for each test suite we want to run with write barrier verification, without requiring custom builds of Unity and mono to enable that. (At least that's the idea, some open questions remain)